### PR TITLE
Added new `class` which handels docker client config and renders it t…

### PR DESCRIPTION
**Pull request by dvonessen**
…o a json string.

What does this PR change? What problem does it solve?
-----------------------------------------------------
The new class helps to handle the creation of docker client configuration within container jobs.
If you need e.g. multiple registries with authentication, you can add them through the class an render it to stdout.

Was the change discussed in an issue?
------------------------------------------------------------

Checklist
---------

- [x] I have read the [Contribution Guidelines](https://github.com/dbsystel/gitlab-ci-python-library/blob/main/CONTRIBUTING.md)
- [x] I have enabled [maintainer edits for this PR](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork)
- [x] I have added tests for all changes in this PR
- [x] I have added api documentation for the changes
- [x] I added a changelog entry to the unreleased section with the PR number. [Changelog](https://github.com/dbsystel/gitlab-ci-python-library/blob/main/CHANGELOG.md)
- [x] Linting is done, and linting jobs are successfull
- [x] I'm done, this Pull Request is ready for review

### CODE REVIEW

File: gcip/addons/container/config.py [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+        Adds a Credentials helper `credHelpers` for a registry.
+        See [docker login#credential-helpers](https://docs.docker.com/engine/reference/commandline/login/#credential-helpers)
+
+        Args:
+            registry (str): Name of the container registry to set `creds_helper` for.
+            cred_helper (str): Name of the credential helper to use together with the `registry`.
+        """
+        compose = {
+            registry: cred_helper
+        }
```
Comment (ThomasSteinbach) [inline gcip/addons/container/config.py:71]: If applying the upper suggestion, then the naming here feels more natural:
```suggestion
        cred_helper = {
            registry: command
        }
```

File: gcip/addons/container/config.py [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+        if "credsHelpers" not in self.config.keys():
+            self.config["credHelpers"] = compose
+        else:
+            self.config["credHelpers"].update(compose)
+
+    def add_auths(
```
Comment (ThomasSteinbach) [inline gcip/addons/container/config.py:77]: You add a single 'auth' to plural 'auths' here:
```suggestion
    def add_auth(
```

File: gcip/addons/container/config.py [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+                "password": "$" + password_env_var
+            }
+        }
+
+        if "auths" not in self.config.keys():
+            self.config["auths"] = compose
+        else:
+            self.config["auths"].update(compose)
+
+    def add_raw(self, raw_input: Dict[Any, Any]) -> None:
```
Comment (ThomasSteinbach) [inline gcip/addons/container/config.py:107]: nice!

File: gcip/addons/container/config.py [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+        Adds arbitrary settings to configuration.
+
+        Be aware and warned! You can overwrite any predefined settings with this method.
+        This method is intendet to be used, if non suitable method is available and you
+        have to set a configuration setting.
+
+        Args:
+            raw_input (Dict[Any, Any]): Dictionary of non-available settings to be set.
+        """
+        self.config.update(raw_input)
```
Comment (ThomasSteinbach) [inline gcip/addons/container/config.py:118]: Should we add a simple helper script for deep merging dicts? Somethins like
https://gist.github.com/angstwad/bf22d1822c38a92ec0a9
? Would not only be useful here.

File: gcip/addons/container/config.py [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+
+    def render(self) -> str:
```
Comment (ThomasSteinbach) [inline gcip/addons/container/config.py:120]:
```suggestion
    def render_create_config_cmd(self) -> str:
```
... oder so ähnlich

File: gcip/addons/container/config.py [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+        """
+        Renders the settings of the docker client config.
```
Comment (ThomasSteinbach) [inline gcip/addons/container/config.py:122]:
```suggestion
        Renders the shell command for creating the docker client config.
```

**Review by ThomasSteinbach — CHANGES_REQUESTED** [review]

File: CHANGELOG.md [hunk 7]
```diff
@@ -15,7 +15,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 * The gcip is now able to detect if two or more jobs would have the same name in the rendered pipeline
   and raises an ValueError to prevent undesirable pipeline behavior.
 * Added new addon to check container optimization with `dive`
-
+* Added new `class` which handels docker client config and renders it to a json string.
```
[the remark this answers is not in the document]
Comment (dvonessen) [reply CHANGELOG.md:18]: Sure.

File: gcip/addons/container/config.py [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+
+        Args:
+            creds_store (str): Should be the suffix of the program to use
+            (i.e. everything after docker-credential-).
+            For example `osxkeychain`, to use docker-credential-osxkeychain or
+            `ecr-login`, to use docker-crendential-ecr-login
+        """
+        self.config["credsStore"] = creds_store
+
+    def add_cred_helper(self, registry: str, cred_helper: str) -> None:
```
[the remark this answers is not in the document]
Comment (dvonessen) [reply gcip/addons/container/config.py:60]: Actually, it is not a command. It is the credentials helper. That is what the documentation says. 
It is more naturally to say "For our `registry` we need to add the `cred_helper`.

File: gcip/addons/container/config.py (hunk 1 above)
Comment (dvonessen) [reply gcip/addons/container/config.py:71]: See comment above

File: gcip/addons/container/config.py (hunk 2 above)
Comment (dvonessen) [reply gcip/addons/container/config.py:77]: Oops, that should have been changed.

File: gcip/addons/container/config.py (hunk 3 above)
Comment (dvonessen) [reply gcip/addons/container/config.py:107]: Thank you!

File: gcip/addons/container/config.py (hunk 6 above)
Comment (dvonessen) [reply gcip/addons/container/config.py:122]: Thx

File: gcip/addons/container/config.py (hunk 5 above)
Comment (dvonessen) [reply gcip/addons/container/config.py:120]: Mhh. What do you think about `get_shell_command()`

File: gcip/addons/container/config.py (hunk 4 above)
Comment (dvonessen) [reply gcip/addons/container/config.py:118]: Mhh yes, would be great. It seems, that the original code has some quirks. e.g. not handling Lists while merging [Comment](https://gist.github.com/angstwad/bf22d1822c38a92ec0a9#gistcomment-2652621). Second could the GPL be a problem? I do not know. The gist is for python2 not python3, there are other contributions in the gist, but as far as I understand neither of them is addressing our problem.
Maybe we should write our own?

File: gcip/addons/container/config.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (ThomasSteinbach) [reply gcip/addons/container/config.py:60]: > Actually, it is not a command.

Isn't every credential helper not a command executed to return credentials?

The official [Docker docs state](https://docs.docker.com/engine/reference/commandline/login/#credential-helpers):

> [the configuration] values specify the suffix of the **program** to use

However `cred_helper` does not sound missleading. You decide.

**Review by ThomasSteinbach — APPROVED** [review]
just replace 'render()' by ...

File: gcip/addons/container/config.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (dvonessen) [reply gcip/addons/container/config.py:60]: As discussed, we will stay with `cred_helper` argument. IMO this is more clear for the user, how this method works.

File: gcip/addons/container/config.py (hunk 4 above)
Comment (dvonessen) [reply gcip/addons/container/config.py:118]: Opened a tracking issue #43

**Pull request merged**